### PR TITLE
Destructuring in variable declarations when module kind is not ES6

### LINKED ISSUE #7074: System module format emits incorrect ES6 object destructuring code
Hello,

I'm using typescript@next: 1.9.0-dev.20160213

When using `target: "es6"` and `module: "system"` configs together, variable object destructuring code like:

```typescript
export var { toString } = 1;
```

or

```typescript
import stuff from 'stuff';
var { toString } = 1;
```

is converted into an incorrect ES6 code:

```javascript
System.register([], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var toString;
    return {
        setters:[],
        execute: function() {
            { toString } = 1; // <-- needs to be wrapped into parentheses to avoid syntax error
        }
    }
});
```

**UPD:** For the case with exported destructured variable

```typescript
export var { toString } = 1;
```

incorrect code is also emitted in all the other module modes due to the same common problem (`var` being dropped, but assignment not wrapped into parentheses to prevent conflict with block syntax):

* `commonjs`:

```javascript
"use strict";
{ exports.toString } = 1;
```

* `amd`:
```javascript
define(["require", "exports"], function (require, exports) {
    "use strict";
    { exports.toString } = 1;
});
```

* `umd`:
```javascript
(function (factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    { exports.toString } = 1;
});
```

### PR DESCRIPTION
**Pull request by vladima**
fixes #7074

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -4218,12 +4218,30 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, ge
...
+                        // emit ES6 destructuring only if target module is ES6 or variable is not exported
+                        // exported variables in CJS\AMD are prefixed with 'exports.' so result javascript { exports.toString } = 1; is illegal
+
+                        const isTopLevelDeclarationInSystemModule =
+                            modulekind === ModuleKind.System &&
+                            shouldHoistVariable(node, /*checkIfSourceFileLevelDecl*/true);
+
+                        if (isTopLevelDeclarationInSystemModule) {
+                            // in System modules top level variables are hoisted 
+                            // so variable declarations with destructuring are turned into destructuring assignments
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: "so variables with destructuring need to be rewritten the same way they would be in ES3/5"

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: This comment seems confusing

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Oh, I see now. What about

```text
So binding patterns will need to be rewritten as expression-level destructuring assignments.
As a result, they will need parentheses to disambiguate object binding patterns from a blocks.
```

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -4218,12 +4218,30 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, ge
...
             function emitVariableDeclaration(node: VariableDeclaration) {
                 if (isBindingPattern(node.name)) {
-                    if (languageVersion < ScriptTarget.ES6) {
-                        emitDestructuring(node, /*isAssignmentExpressionStatement*/ false);
-                    }
-                    else {
+                    const isExported = getCombinedNodeFlags(node) & NodeFlags.Export;
+                    if (languageVersion >= ScriptTarget.ES6 && (!isExported || modulekind === ModuleKind.ES6)) {
+                        // emit ES6 destructuring only if target module is ES6 or variable is not exported
+                        // exported variables in CJS\AMD are prefixed with 'exports.' so result javascript { exports.toString } = 1; is illegal
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: /

File: src/compiler/emitter.ts (hunk 1 above)
Comment (vladima) [inline src/compiler/emitter.ts]: :white_check_mark:

File: src/compiler/emitter.ts (hunk 2 above)
Comment (vladima) [inline src/compiler/emitter.ts]: :white_check_mark:

**Comment by RyanCavanaugh** [comment]
:+1:

**Comment by mhegazy** [comment]
:+:

**Pull request merged**